src/main/java/org/terasology/taskSystem/actions/PerformTaskNode.java

```diff
@@ -41,7 +47,18 @@
     private static final Logger logger = LoggerFactory.getLogger(PerformTaskNode.class);
 
     @In
-    EntityManager entityManager;
+    private Context context;
+
+    private PlantTaskCompletionSystem plantTaskCompletion;
```

Comment:
I implemented these as systems so we do not have to create a new instance of the task completion classes with every task.

Reply:
There's nothing wrong with creating an instance of a class, running a method on it, and then letting it be garbage collected.   This is cheap operation.   This happens every time you do something like `String x = "first" + "last";`

Keeping the class around in memory as part of the Context has a higher eventual cost.  It will never be garbage collected.   It will be one more item that has to be searched through when pulling objects out of the data context.    There is also the possibility of multithreading issues if you add an instance variable to your class and reuse it for several oreons.   That's not currently the case, though.

Note that there is not a huge cost difference either way.

Another approach if you do want to have a single class which persists is to declare the variable without @In inside performTaskNode and create the class the first time it is used in that instance variable.    Have a getPlantTaskCompletion() method which creates a new PlantTaskCompletion object if the instance variable plantTaskCompletion is null.

```
private PlantTaskCompletion getPlantTaskCompletion() {
    if (null == plantTaskCompletion) {
        plantTaskCompletion = new PlantTaskCompletion();
    }

    return plantTaskCompletion;
}
```

My recommendation is that systems be saved for API points.    Use a system if there are system methods which would eventually be accessed by another developer.

I would also rename this to PlantingTask rather than PlantTaskCompletion.   I would put all aspects of working with a Planting task into that object.   Same with other tasks.

Finally, I would pass into PlantingTask the variables that you are currently using as instance variables (blockManager, blockEntityRegistry).   If you start writing unit tests at some point, this will make it easier to test your code as well as keep your code safe for multithreading.

```
    plantTask.placeCrops(blockManager, blockEntityRegistry, selectedRegion, Constants.OREON_CROP_PREFAB);
```